### PR TITLE
Better error messages

### PR DESCRIPTION
**Pull request by EntropyFails**
I've been having some fun playing with cooltt. An annoyance I had is the opaque error message on parse error. So I added the next token to the LexingUtil. Because the span is used in the Grammar I needed to change the call to pp_span to take in a product type. 

More controversially, if the output of cooltt is going to be used by another program, the INFO line numbers are a bit of a problem so I suppressed them in pp_message.

My OCaml is meh so happy to adjust this pull request to however you'd prefer it.

### CODE REVIEW

File: src/lib/Log.ml [hunk 1]
```diff
@@ -13,12 +13,22 @@ let pp_lvl fmt =
...
   | None ->
     pp Format.std_formatter data
   | Some span ->
-    Format.printf "@[<v>%a [%a]:@,  @[<v>%a@]@]@.@."
-      LexingUtil.pp_span span
-      pp_lvl lvl
-      pp data
+    match lvl with
+    | `Info ->
+      Format.printf "@[<v>%a@]@."
```
Comment (jonsterling) [inline src/lib/Log.ml:23]: I would prefer to keep the level and span, because it is part of the protocol that allows cooltt to be used in our editor mode; the purpose is to display an indicator near where the information emanated.

File: src/basis/LexingUtil.ml [hunk 2]
```diff
@@ -4,11 +4,12 @@ type span =
   {start : Lexing.position;
    stop : Lexing.position}
 
-let pp_span : span Pp.printer =
-  fun fmt span ->
-  Format.fprintf fmt "%a:%i.%i-%i.%i"
+let pp_span : (span * string) Pp.printer =
+  fun fmt (span,last_token) ->
+  Format.fprintf fmt "%a:%i.%i-%i.%i near %s"
```
Comment (jonsterling) [inline src/basis/LexingUtil.ml:9]: Perhaps the last token should not be printed with the span, but instead should be part of the `ParseError` message etc. Then, the span output protocol can remain the same, but we can report "There was a parsing error near [token]" or something.

**Comment by favonia** [comment]
> More controversially, if the output of cooltt is going to be used by another program, the INFO line numbers are a bit of a problem so I suppressed them in pp_message.

I want to learn more about the use cases. It is indeed in the plan to have (separate) scriptable interactive interfaces. Could you elaborate the usage in your mind? Thanks! (And thanks for your contribution.)

File: src/lib/Log.ml (hunk 1 above)
Comment (EntropyFails) [reply src/lib/Log.ml]: Understood. I've uploaded a change that preserves the formatting.

File: src/basis/LexingUtil.ml (hunk 2 above)
Comment (EntropyFails) [reply src/basis/LexingUtil.ml]: I've changed it in the latest commit to keep the info line the same and just add error messages near the pp data line.

**Comment by EntropyFails** [comment]
> > More controversially, if the output of cooltt is going to be used by another program, the INFO line numbers are a bit of a problem so I suppressed them in pp_message.
> 
> I want to learn more about the use cases. It is in the plan to have (separate) scriptable interactive interfaces. Could you elaborate the usage in your mind? Thanks! (And thanks for your contribution.)

Eventually, if this gets beyond the airy lands of research, someone will want to take values it outputs and use them in the real world. It kind of reminds me of the early days of Haskell before Monads where all you could do is output a string to do IO. =) 

Another way of doing this that doesn't alter your current toolset is to provide a command line argument that changes how the pretty printer works if you want "cpu consumable output".

File: src/lib/Monads.ml [hunk 3]
```diff
@@ -267,7 +267,7 @@ struct
 
   let emit ?(lvl = `Info) loc pp a : unit m =
     fun (st, _env) ->
-    Log.pp_message ~loc ~lvl pp a;
+    Log.pp_message ~loc ~lvl ~last_token:"" pp a;
```
Comment (EntropyFails) [inline src/lib/Monads.ml:270]: I'm too scared to change emit at the moment because I don't understand the refiner... =)

**Comment by favonia** [comment]
> Eventually, if this gets beyond the airy lands of research, someone will want to take values it outputs and use them in the real world. It kind of reminds me of the early days of Haskell before Monads where all you could do is output a string to do IO. =)
> 
> Another way of doing this that doesn't alter your current toolset is to provide a command line argument that changes how the pretty printer works if you want "cpu consumable output".

Yes, this is exactly what I had in my mind. I haven't thought through it very carefully, though. Apparently, we need to write some code to un-elaborate core terms back to surface syntax. Assuming there's no problem there, I wonder if you have a design you like in your mind? (e.g., a tool that outputs exactly what you want.) I believe it's not difficult to add a new output mode.

**Comment by jonsterling** [comment]
I will re-review this hopefully today, thank you again for your contribution! We are very excited to see some outside interest, and would be happy to discuss ways to further involve you in the project if you have interest.

**Comment by EntropyFails** [comment]
Give me a second. I better understand the difference between load_file and the REPL now and I'll have a patch shortly that makes this better.

File: src/lib/DriverMessage.ml [hunk 4]
```diff
@@ -0,0 +1,6 @@
+type message =
```
Comment (EntropyFails) [inline src/lib/DriverMessage.ml:1]: It is required to pull this out into it's own module to avoid a dependency cycle between the Driver and Conversions

**Comment by EntropyFails** [comment]
I like this better as it starts to abstract the difference between the an error and output. However don't review just yet as I found an edge case...

**Comment by EntropyFails** [comment]
I like this better as it starts the process of abstracting Error Messaging from Eval messaging. The test suite runs and my testing of various errors works. 
The errors currently look like...
```
my-test/errortest.cooltt:37.4-37.5 [Error]:
  [Error near normalize] Parse error
```

@favonia I haven't thought super deeply about it either. But my intuition says that if we provide a system for taking in strings and outputting strings made for machine use, we'll get the benefits of Unix style integrations out of the gate and that could lead to some surprising things.

For me, I just want a language where a + b can be known to equal b + a but in a "turtles all the way down" way. I see the same world where someone comes up with a cool proof of some property of a structure that my program is using, I add it and my program is "magically" more expressive as you all see so I'm happy to chip in some to push that along. =)

File: src/lib/Load.mli [hunk 5]
```diff
@@ -1,4 +1,4 @@
-type error = LexingError of Basis.LexingUtil.span | ParseError of Basis.LexingUtil.span
+type error = LexingError of Basis.LexingUtil.span * string | ParseError of Basis.LexingUtil.span * string
```
Comment (jonsterling) [inline src/lib/Load.mli:1]: I suggest changing this to be an inline record so that the intention of the `string` thing is clear from the interface. We could write something like:

```ocaml
type error = 
| LexingError of {span : ...; last_token : string}
| ...
```

File: src/lib/DriverError.ml [hunk 6]
```diff
@@ -0,0 +1,2 @@
+
+type error_with_maybe_span =  ErrorWithMaybeSpan of DriverMessage.message * Basis.LexingUtil.span option
```
Comment (jonsterling) [inline src/lib/DriverError.ml:2]: I think we can make this a little more idiomatic, like:

```ocaml
type error = {message : DriverMessage.message; span : Basis.LexingUtil.span option}
```

File: src/lib/Log.ml [hunk 7]
```diff
@@ -2,7 +2,7 @@ open Basis
 
 type level = [`Info | `Error | `Warn]
 
-type location = LexingUtil.span option
+type location = DriverError.error_with_maybe_span option
```
Comment (jonsterling) [inline src/lib/Log.ml:5]: This interface is kind of strange, because `pp_message` is now taking an error, but `pp_message` isn't always about errors... Perhaps I am getting hung up on the names of things.

Perhaps it would be helpful if you could think a bit about how you want all this stuff to work, and then summarize the interface you favor?

Comment (EntropyFails) [reply src/lib/Log.ml]: Log only ever outputs from Driver.process_file... that only ever outputs errors.

File: src/lib/DriverError.ml (hunk 6 above)
Comment (EntropyFails) [reply src/lib/DriverError.ml]: I changed this up to be better by removing DriverError entirely and making a more clear DriverMessage module

File: src/lib/Load.mli (hunk 5 above)
Comment (EntropyFails) [reply src/lib/Load.mli]: I cleared this up in the latest version.

**Comment by EntropyFails** [comment]
Agree with your comments... this latest is further along the way by maintaining one DriverMessage class, splitting the load errors from runtime errors, and taking Driver Messaging in pp_message out of the Driver file and putting it into the DriverMessaging module.

There is always more to do but this seems a lot further towards "better". Happy for feedback!

**Comment by jonsterling** [comment]
So, this is looking a lot better! There are still some aspects that I am finding confusing at the moment; the part I'm still getting stuck on is there is a type called `error` that contains non-error things (by virtue of containing `DriverMessage.message`), and consequently, the function `pp_error_message` is therefore also seeming to be printing non-error things. Maybe the right question for me to ask is: perhaps the errors (like `ParseError` etc.) should not be part of `DriverMessage.message` and some types should be peeled apart?

I am also finding the distribution of the spans and the last-tokens to be a little confusing; there is some semi-subtle layering going on in the architecture and I could use a high-level gloss of the motivations.

Thanks again!

**Comment by EntropyFails** [comment]
I definitely see where you are going with your comment and I agree. I'll move around some things and try to get closer to cleanly separating messaging from error reporting.

I've taken a look at how Haskell does it and instead of having a LexingUtil class, they have an [Outputable](https://gitlab.haskell.org/ghc/ghc/-/blob/d4a0be758003f32b9d9d89cfd14b9839ac002f4d/compiler/GHC/Utils/Outputable.hs) class that is used by an [Error](https://gitlab.haskell.org/ghc/ghc/-/blob/d4a0be758003f32b9d9d89cfd14b9839ac002f4d/compiler/GHC/Tc/Errors.hs) and the main [GHC](https://gitlab.haskell.org/ghc/ghc/-/blob/d4a0be758003f32b9d9d89cfd14b9839ac002f4d/compiler/GHC.hs) class. That is obviously better but I won't get to that level of sophistication in this PR. Let me propose something incrementally better as I walk towards that goal.

File: src/lib/DriverMessage.mli [hunk 8]
```diff
@@ -0,0 +1,15 @@
+(* Driver Messages come in 2 types, either an error message or a normalized term/Definition*)
```
Comment (jonsterling) [inline src/lib/DriverMessage.mli:1]: I like the new design of this module!

Comment (EntropyFails) [reply src/lib/DriverMessage.mli:1]: Thanks! Yeah I was misunderstanding where the Syntax.pp was being used so I avoided adding more stuff to data. Now that I have a better handle on it, the latest patch should be much cleaner.
